index/segment/fs/encoding/fields/index.go

```diff
@@ -0,0 +1,145 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package fields
+
+import (
+	"fmt"
+	"io"
+
+	"github.com/m3db/m3ninx/index/segment/fs/encoding"
+	"github.com/m3db/m3ninx/postings"
+)
+
+const (
+	indexHeaderSize = 8 // Base postings ID as a uint64.
+
+	initialIndexEncoderLen     = 1024
+	indexEncoderWriteThreshold = 512
+)
+
+// IndexWriter is a writer for the index file for stored fields.
+type IndexWriter struct {
+	writer io.Writer
+	enc    *encoding.Encoder
+	ready  bool
+	prev   postings.ID
+}
+
+// NewIndexWriter returns a new IndexWriter.
+func NewIndexWriter(w io.Writer) *IndexWriter {
+	iw := &IndexWriter{
+		writer: w,
+		enc:    encoding.NewEncoder(initialIndexEncoderLen),
+	}
+	return iw
+}
+
+func (w *IndexWriter) Write(id postings.ID, offset uint64) error {
+	if !w.ready {
+		w.enc.PutUint64(uint64(id))
+		w.ready = true
+	} else {
+		if id != w.prev+1 {
+			return fmt.Errorf("non-sequential postings ID: receieved %v but previous ID was %v", id, w.prev)
+		}
+	}
+
+	w.enc.PutUint64(offset)
+	w.prev = id
+
+	if w.enc.Len() > indexEncoderWriteThreshold {
+		if err := w.write(); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (w *IndexWriter) write() error {
+	b := w.enc.Bytes()
+	n, err := w.writer.Write(b)
+	if err != nil {
+		return err
+	}
+	if n < len(b) {
+		return io.ErrShortWrite
+	}
+	w.enc.Reset()
+	return nil
+}
+
+// Close closes the IndexWriter and ensures any buffered data is written out.
+func (w *IndexWriter) Close() error {
+	if w.enc.Len() == 0 {
+		return nil
+	}
+	return w.write()
+}
+
+// IndexReader is a reader for the index file for stored fields.
+type IndexReader struct {
+	data  []byte
+	dec   *encoding.Decoder
+	base  postings.ID
+	limit postings.ID
+}
+
+// NewIndexReader returns a new IndexReader.
+func NewIndexReader(data []byte) (*IndexReader, error) {
+	if len(data) < indexHeaderSize {
+		return nil, io.ErrShortBuffer
+	}
+
+	payloadLen := len(data) - indexHeaderSize
+	if payloadLen%8 != 0 {
+		return nil, fmt.Errorf("stored fields index payload should be a multiple of 8, found %v", payloadLen%8)
+	}
+	count := payloadLen / 8
```

Comment:
hm, don't you need to remove the base id from these calcs?

Reply:
That's done in line 112 when we subtract `indexHeaderSize `, right now the header only contains the base ID.

Reply:
Yeah the var name confused me. maybe change to indicate it's the base id?